ding/policy/ibc.py

```diff
@@ -0,0 +1,185 @@
+            obs, action = data['obs'], data['action']
+            if len(obs.shape) == 1:
+                obs = obs.unsqueeze(-1)
+            if len(action.shape) == 1:
+                action = action.unsqueeze(-1)
+
+            # (B, N, O), (B, N, A)
+            obs, negatives = self._stochastic_optimizer.sample(obs, self._learn_model)
+
+            # (B, N+1, A)
+            targets = torch.cat([action.unsqueeze(dim=1), negatives], dim=1)
+            # (B, N+1, O)
+            obs = torch.cat([obs[:, :1], obs], dim=1)
+
+            permutation = torch.rand(targets.shape[0], targets.shape[1]).argsort(dim=1)
+            targets = targets[torch.arange(targets.shape[0]).unsqueeze(-1), permutation]
+
+            # (B, )
+            ground_truth = (permutation == 0).nonzero()[:, 1].to(self._device)
+
+            # (B, N+1) for ebm
+            # (B, N+1, A) for autoregressive ebm
+            energy = self._learn_model.forward(obs, targets)
+
+            logits = -1.0 * energy
+            if isinstance(self._stochastic_optimizer, AutoRegressiveDFO):
+                # autoregressive case
+                # (B, A)
+                ground_truth = unsqueeze_repeat(ground_truth, logits.shape[-1], -1)
+            loss = F.cross_entropy(logits, ground_truth)
+            loss_dict['ebm_loss'] = loss.item()
+
+            if isinstance(self._stochastic_optimizer, MCMC):
+                grad_penalty = self._stochastic_optimizer.grad_penalty(obs, targets, self._learn_model)
+                loss += grad_penalty
+                loss_dict['grad_penalty'] = grad_penalty.item()
+            loss_dict['total_loss'] = loss.item()
+
+            self._optimizer.zero_grad()
+            loss.backward()
+            with self._sync_timer:
+                if self._cfg.learn.multi_gpu:
+                    self.sync_gradients(self._learn_model)
+            sync_time = self._sync_timer.value
+            self._optimizer.step()
+
+        total_time = self._timer.value
+
+        return {
+            'total_time': total_time,
+            'sync_time': sync_time,
+            **loss_dict,
+        }
+
+    def _monitor_vars_learn(self):
+        if isinstance(self._stochastic_optimizer, MCMC):
+            return ['total_loss', 'ebm_loss', 'grad_penalty', 'total_time', 'sync_time']
+        else:
+            return ['total_loss', 'ebm_loss', 'total_time', 'sync_time']
+
+    def _init_eval(self):
+        self._eval_model = model_wrap(self._model, wrapper_name='base')
+        self._eval_model.reset()
+
+    def _forward_eval(self, data: dict) -> dict:
+        tensor_input = isinstance(data, torch.Tensor)
+        if not tensor_input:
+            data_id = list(data.keys())
+            data = default_collate(list(data.values()))
+
+        if self._cuda:
+            data = to_device(data, self._device)
+
+        self._eval_model.eval()
+        with torch.set_grad_enabled(isinstance(self._stochastic_optimizer, MCMC)):
```

Comment:
can we move it into `stochastic_optimizer.infer` method

Reply:
Actually, we don't need this line cuz I wrap `MCMC.infer` with `@no_ebm_grad()` and `DFO/ARDFO.infer` with `@torch.no_grad()`.  I just have it here so that others can see the difference between MCMC and DFO from the `ibc.py` without having to check the details of `ebm.py`

Reply:
This line is actually redundant.